### PR TITLE
added support for OvirtError, for making things more unified for versions

### PR DESCRIPTION
**Pull request by ik5**

### CODE REVIEW

**Comment by ik5** [comment]
@orrabin This is the first task we talked about, when it is accepted, I will add the supported exceptions to Foreman as well.

**Comment by orrabin** [comment]
@ik5 This adds a new ovirt error but it does not replace the existing OVIRT::OvirtException.
Any error we receive from rbovirt will still get to core as an OvirtException: https://github.com/abenari/rbovirt/blob/995176066ad3d6f6d29328e8ee4f0a98a02d300e/lib/rbovirt.rb#L39

**Comment by ik5** [comment]
Oh, so I'm going to make it use the new exception.

**Comment by ik5** [comment]
@orrabin Now it captures all of the `OvirtException` when the `client` is called, and raising `Fog::Ovirt::Errors::OvirtError` instead.

Oh, and sorry, but there were two or three places where my OCD kicked in, and saw something like `def foo arg` that fixed to `def foo(arg)`, that's because it's something that really drives me nuts, and can't ignore, sorry for advance.

File: lib/fog/ovirt/models/compute/template.rb [hunk 1]
```diff
@@ -42,11 +42,15 @@ def ready?
...
+        rescue ::OVIRT::OvirtException => e
+          raise ::Fog::Ovirt::OvertError, e.message
         end
 
         def save
-          raise Fog::Errors::Error.new('Resaving an existing object may create a duplicate') if persisted?
+          raise ::Fog::Ovirt::Errors::OvirtError, 'Providing an existing object may create a duplicate object' if persisted?
           service.client.create_template(attributes)
+        rescue ::OVIRT::OvirtException => e
+          raise ::Fog::Ovirt::OvertError, e.message
```
Comment (ShimShtein) [inline lib/fog/ovirt/models/compute/template.rb]: Typo, `ovErtError` instead of `Ovirt`

File: lib/fog/ovirt/requests/compute/activate_volume.rb [hunk 2]
```diff
@@ -7,6 +7,9 @@ def activate_volume(id, options)
           raise ArgumentError, "volume id is a required parameter for activate-volume" unless options.key? :id
 
           client.activate_volume(id, options[:id])
+
+        rescue ::OVIRT::OvirtException => e
+          raise ::Fog::Ovirt::OvertError, e.message
```
Comment (ShimShtein) [inline lib/fog/ovirt/requests/compute/activate_volume.rb]: Why are we sending only the message? I would like to see `::Fog::Ovirt::OvertError` wrapping the complete `OvirtException` instance, so I will be able to see the real source of the error.

File: lib/fog/ovirt/models/compute/template.rb [hunk 3]
```diff
@@ -42,11 +42,15 @@ def ready?
...
           service.client.destroy_template(id)
+        rescue ::OVIRT::OvirtException => e
+          raise ::Fog::Ovirt::OvertError, e.message
         end
 
         def save
-          raise Fog::Errors::Error.new('Resaving an existing object may create a duplicate') if persisted?
+          raise ::Fog::Ovirt::Errors::OvirtError, 'Providing an existing object may create a duplicate object' if persisted?
           service.client.create_template(attributes)
+        rescue ::OVIRT::OvirtException => e
```
Comment (ShimShtein) [inline lib/fog/ovirt/models/compute/template.rb]: Can we wrap the `client` with a generic exception handler? It would be nicer, and future client calls would be covered automatically.

I mean something like:
``` ruby
class ClientWrapper
  def initialize(client)
     @actual_client = client
  end

  def method_missing(method, *params)
    @actual_clent.send(method, *params)
    rescue ::OVIRT::OvirtException => e
      raise ::Fog::Ovirt::OvirtError, e
    end
  end
end
```

**Review by ShimShtein — CHANGES_REQUESTED** [review]

File: lib/fog/ovirt/models/compute/template.rb (hunk 3 above)
Comment (ik5) [reply lib/fog/ovirt/models/compute/template.rb]: At the moment it does not work like that.
there is a definition of `@client` that is an ovirt sdk client, and a method that does a read only property for it named `client`.

Each place I went to, opens that specific class, and add a client.
If I understand what you are suggesting, it means to rewrite the whole concept of how it works in order to have the same way of doing stuff. But that's task number two, to my recollection.

File: lib/fog/ovirt/models/compute/template.rb (hunk 1 above)
Comment (ik5) [reply lib/fog/ovirt/models/compute/template.rb]: fixed

File: lib/fog/ovirt/requests/compute/activate_volume.rb (hunk 2 above)
Comment (ik5) [reply lib/fog/ovirt/requests/compute/activate_volume.rb]: done.

File: lib/fog/ovirt/models/compute/template.rb (hunk 3 above)
Comment (ShimShtein) [reply lib/fog/ovirt/models/compute/template.rb]: My bad, this comment belongs to request class, not the model. I am talking about wrapping the [client variable](https://github.com/ik5/fog-ovirt/blob/8c311047a9e597fa29a1614196891291454b0ab6/lib/fog/ovirt/compute.rb#L143) in the request, so calls to client won't emit `::OVIRT::OvirtException` at all.

File: lib/fog/ovirt/models/compute/template.rb (hunk 1 above)
Comment (ShimShtein) [reply lib/fog/ovirt/models/compute/template.rb]: Could you add some tests to this? It looks like nobody covered that code path.

File: lib/fog/ovirt/models/compute/template.rb (hunk 3 above)
Comment (ik5) [reply lib/fog/ovirt/models/compute/template.rb]: done

**Comment by ik5** [comment]
@orrabin  I'm unable to test it myself, can you give me a tip on how to it properly?

File: lib/fog/ovirt/core.rb [hunk 4]
```diff
@@ -10,6 +10,7 @@ module Errors
       class ServiceError < Fog::Errors::Error; end
       class SecurityError < ServiceError; end
       class NotFound < ServiceError; end
+      class OvirtError < Fog::Errors::Error; end
```
Comment (ShimShtein) [inline lib/fog/ovirt/core.rb:13]: Why not splitting errors to fog-generated base and derived wrapper exception that will save the original exception object:

``` ruby
class OvirtError < Fog::Errors::Error; end

class OvirtClientError < OvirtError
  attr_reader :original
  def initialize(client_exception)
    @original = client_exception
    super("Ovirt client returned an error: #{client_exception.message}")
  end
end
```

File: lib/fog/ovirt/models/compute/template.rb [hunk 5]
```diff
@@ -42,10 +42,12 @@ def ready?
 
         def destroy(options = {})
           service.client.destroy_template(id)
+        rescue ::OVIRT::OvirtException => e
```
Comment (ShimShtein) [inline lib/fog/ovirt/models/compute/template.rb]: If the client is already wrapped, this should be unnecessary.

**Review by ShimShtein — CHANGES_REQUESTED** [review]
Thanks for reformatting, a couple of comments though.

File: lib/fog/ovirt/models/compute/template.rb (hunk 5 above)
Comment (ik5) [reply lib/fog/ovirt/models/compute/template.rb]: removed it, didn't find any other locations beside `method_missing`

File: lib/fog/ovirt/core.rb (hunk 4 above)
Comment (ik5) [reply lib/fog/ovirt/core.rb:13]: done

File: lib/fog/ovirt/compute.rb [hunk 6]
```diff
@@ -101,6 +101,26 @@ def ovirt_attrs obj
...
       end
 
+      class ClientWrapper
+        def initialize(client)
+          @client = client
+        end
+
+        def method_missing(symbol, *args, &block)
+          if block_given?
+            yield
```
Comment (ShimShtein) [inline lib/fog/ovirt/compute.rb]: Shouldn't we just pass the block to real client?

File: lib/fog/ovirt/compute.rb [hunk 7]
```diff
@@ -101,6 +101,26 @@ def ovirt_attrs obj
         end
       end
 
+      class ClientWrapper
```
Comment (ShimShtein) [inline lib/fog/ovirt/compute.rb]: Can you rename it to something that tells us what it is doing, `ExceptionWrapper` for example?

File: lib/fog/ovirt/compute.rb [hunk 8]
```diff
@@ -140,18 +160,18 @@ def initialize(options={})
           connection_opts[:ca_no_verify]  = options[:ovirt_ca_no_verify]
           connection_opts[:filtered_api]  = options[:ovirt_filtered_api]
 
-          @client = OVIRT::Client.new(username, password, url, connection_opts)
+          @client = ClientWrapper.new(OVIRT::Client.new(username, password, url, connection_opts))
         end
 
         def api_version
           client.api_version
+        rescue ::OVIRT::OvirtException => e
```
Comment (ShimShtein) [inline lib/fog/ovirt/compute.rb]: No need for this, we already do it in the `client`

File: lib/fog/ovirt/core.rb [hunk 9]
```diff
@@ -10,6 +10,16 @@ module Errors
...
+
+      class OvirtClientError < OvirtError
+        attr_reader :orign_exception
```
Comment (ShimShtein) [inline lib/fog/ovirt/core.rb]: Typo? `:origN_exception` vs `@orig_exception`

**Review by ShimShtein — CHANGES_REQUESTED** [review]
Please, pretty please, can you add `shindo` tests for all this new functionality?

File: lib/fog/ovirt/core.rb (hunk 9 above)
Comment (ik5) [reply lib/fog/ovirt/core.rb]: good catch

File: lib/fog/ovirt/compute.rb (hunk 8 above)
Comment (ik5) [reply lib/fog/ovirt/compute.rb]: opps, missed that point, sorry

File: lib/fog/ovirt/compute.rb (hunk 6 above)
Comment (ik5) [reply lib/fog/ovirt/compute.rb]: not sure, because the block/proc part will be belonging to the wrapper, for something to do. Not sure if it should be sent to the real client.

File: lib/fog/ovirt/compute.rb (hunk 7 above)
Comment (ik5) [reply lib/fog/ovirt/compute.rb]: done

**Comment by ik5** [comment]
@ShimShtein, still trying to figure out how to test it.
The Mock part ignores the client completely, and trying to see how to make a mock that does use it, so I can test it.

File: lib/fog/ovirt/compute.rb (hunk 6 above)
Comment (ShimShtein) [reply lib/fog/ovirt/compute.rb]: The binding of the block is to the caller, but I think that it is intended to run in the context of the client anyway. It can really be a problem, consider:

``` ruby
# rbovirt client implementation:
class Client
  def inner_throw_method
    throw ::OVIRT::OvirtException
  end

  def outer_method
    puts 'this should run first'
    yield
  end
end

# Fog request implementation
class FogOvirtRequest
  def do_something
    client.outer_method do
      inner_throw_method
    rescue ::OVIRT::OvirtException
      log 'This is expected' 
    end
  end
end
```

It will fail with the your suggested implementation for two reasons: the exception would be wrapped and the order will be changed.

Comment (ik5) [reply lib/fog/ovirt/compute.rb]: i see your point there, you are right, changing it

**Comment by ik5** [comment]
@ShimShtein here are the tests :)

File: tests/ovirt/requests/compute/client_tests.rb [hunk 10]
```diff
@@ -0,0 +1,16 @@
...
+    raise OVIRT::OvirtException, 'Test'
+  end
+
+  begin
+    object_under_test.foo
+  rescue Fog::Ovirt::Errors::OvirtClientError => e
+    error = e
+  end
+
+  tests('Has valid backtrace').returns(error.nil?) { false }
```
Comment (ShimShtein) [inline tests/ovirt/requests/compute/client_tests.rb]: Almost there :smile: 
I think we can skip the backtrace test altogether, instead, can you please test that:

1. Exception properly wraps the text: `error.message == 'Ovirt client returned an error: '`
2. Exception includes the original erorr: `error.orig_exception.is_a?(OVIRT::OvirtException) && error.orig_exception.message == 'Test'`

Comment (ik5) [reply tests/ovirt/requests/compute/client_tests.rb]: The second test exists, but forgot to commit it (sorry).

I think that your suggestion for the first test is actually much better then what I was thinking in doing :)

**Comment by ik5** [comment]
following some shindo examples, found a nicer way to do it :)

File: tests/ovirt/requests/compute/client_tests.rb [hunk 11]
```diff
@@ -0,0 +1,23 @@
...
+    @object_under_test.foo
+  end
+
+  tests('The exception test is as expected').returns(true) do
+    begin
+      @object_under_test.foo
+    rescue Fog::Ovirt::Errors::OvirtClientError => e
+      e.message == 'Ovirt client returned an error: Test'
+    end
+  end
```
Comment (ShimShtein) [inline tests/ovirt/requests/compute/client_tests.rb]: Can you also test that you get the original exception in `orig_exception` (2 in my previous comment)?

Comment (ik5) [reply tests/ovirt/requests/compute/client_tests.rb]: sorry, my bad, here it is :)

**Review by ShimShtein — APPROVED** [review]
:+1:  ACK from me.

**Comment by orrabin** [comment]
@ik5 please squash the commits

File: lib/fog/ovirt/version.rb [hunk 12]
```diff
@@ -1,5 +1,5 @@
 module Fog
   module Ovirt
-    VERSION = '0.1.2'
+    VERSION = '0.1.3'.freeze
```
Comment (orrabin) [inline lib/fog/ovirt/version.rb]: why the version bump? why freeze?

**Review by orrabin — COMMENTED** [review]
Tested by:
- debugging the `test_connection` method does get the correct message with this pr:
``` #<Fog::Ovirt::Errors::OvirtClientError: Ovirt client returned an error: 401 Unauthorized>```
- trying to save a vm with no cores or memory returns ` Fog::Ovirt::Errors::OvirtClientError`

File: lib/fog/ovirt/version.rb (hunk 12 above)
Comment (ik5) [reply lib/fog/ovirt/version.rb]: > why the version bump? 

It changes how exception works, so I think it should be a new version, so exception handlers will need to be changed.

> why freeze?

Ruby's constant are mutable by defaults. Moreover, each time you call a constant string, it allocate new instance of that string.
Calling `freeze` does two main things here:
1. Makes the constant immutable (more or less)
2. Use the constant allocation once, rather then allocate new memory each usage.

**Comment by ik5** [comment]
I'ved squashed everything, and no version bump

File: lib/fog/ovirt/version.rb [hunk 13]
```diff
@@ -1,5 +1,5 @@
 module Fog
   module Ovirt
-    VERSION = '0.1.2'
+    VERSION = '0.1.2'.freeze
```
Comment (orrabin) [inline lib/fog/ovirt/version.rb]: Please move this to a separate pr, there are now merge conflicts because the version was bumped before this pr was merged.

Comment (ik5) [reply lib/fog/ovirt/version.rb]: sorry on that :'( removed the freeze

Comment (orrabin) [reply lib/fog/ovirt/version.rb]: The freeze was a good change, just didn't want it to affect the rest of the pr.
It's already suggested in: https://github.com/fog/fog-ovirt/pull/5/files#diff-73b69b006fbdd314071398980a75b3eaR3

File: lib/fog/ovirt/core.rb [hunk 14]
```diff
@@ -10,6 +10,16 @@ module Errors
       class ServiceError < Fog::Errors::Error; end
       class SecurityError < ServiceError; end
       class NotFound < ServiceError; end
+      class OvirtError < Fog::Errors::Error; end
+
+      class OvirtClientError < OvirtError
```
Comment (orrabin) [inline lib/fog/ovirt/core.rb]: This is a little confusing, `OvirtClient` does not show this is an actual error we received from Ovirt.
Maybe this can be renamed to `OvirtEngineError`?
My first impression was that `OvirtError` came from ovirt so I didn't understand it was the base class.

Comment (ik5) [reply lib/fog/ovirt/core.rb]: here is a rename, tested it as well.

**Review by orrabin — APPROVED** [review]
LGTM

**Comment by ShimShtein** [comment]
Merged, thanks @ik5!
